src/Evolution/Systems/Cce/PrecomputeCceDependencies.hpp

```diff
@@ -0,0 +1,261 @@
+                     Spectral::Swsh::Tags::derivative_spin_weight<DerivKind>>*>
+        d_r_divided_by_r,
+    const SpinWeighted<ComplexDataVector, 0>& boundary_r, size_t l_max,
+    size_t number_of_radial_points) noexcept;
+
+}  // namespace detail
+
+/*!
+ * \brief A set of procedures for computing the set of inputs to the CCE
+ * integrand computations that can be computed before any of the intermediate
+ * integrands are evaluated.
+ *
+ * \details The template specializations of this template are
+ * compatible with acting as a the mutator in a \ref DataBoxGroup
+ * `db::mutate_apply` operation. For flexibility in defining the \ref
+ * DataBoxGroup structure, the tags for `Tensor`s used in these functions are
+ * also organized into type lists:
+ * -  type alias `integration_independent_tags`: with a subset of
+ * `Cce::pre_computation_tags`, used for both input and output.
+ * -  type alias `boundary_values`: with a subset of
+ *   `Cce::pre_computation_boundary_tags`, used only for input.
+ * - type alias `pre_swsh_derivatives` containing hypersurface quantities. For
+ * this struct, it will only ever contain `Cce::Tags::BondiJ`, and is used as
+ * input.
+ *
+ * The `BoundaryPrefix` tag allows easy switching between the
+ * regularity-preserving version and standard CCE
+ *
+ */
+template <template <typename> class BoundaryPrefix, typename Tag>
+struct PrecomputeCceDependencies;
+
+/// Computes \f$1 - y\f$ for the Cce system.
+template <template <typename> class BoundaryPrefix>
+struct PrecomputeCceDependencies<BoundaryPrefix, Tags::OneMinusY> {
+  using boundary_tags = tmpl::list<>;
+  using pre_swsh_derivative_tags = tmpl::list<>;
+  using integration_independent_tags = tmpl::list<>;
+
+  using return_tags = tmpl::list<Tags::OneMinusY>;
+  using argument_tags = tmpl::list<Spectral::Swsh::Tags::LMax,
+                                   Spectral::Swsh::Tags::NumberOfRadialPoints>;
+
+  static void apply(
+      const gsl::not_null<Scalar<SpinWeighted<ComplexDataVector, 0>>*>
+          one_minus_y,
+      const size_t l_max, const size_t number_of_radial_points) noexcept {
+    const size_t number_of_angular_points =
+        Spectral::Swsh::number_of_swsh_collocation_points(l_max);
+    const auto& one_minus_y_collocation =
+        1.0 - Spectral::collocation_points<Spectral::Basis::Legendre,
+                                           Spectral::Quadrature::GaussLobatto>(
+                  number_of_radial_points);
+    // iterate through the angular 'chunks' and set them to their 1-y value
+    for (size_t i = 0; i < number_of_radial_points; ++i) {
+      ComplexDataVector angular_view{
+          get(*one_minus_y).data().data() + number_of_angular_points * i,
+          number_of_angular_points};
+      angular_view = one_minus_y_collocation[i];
+    }
+  }
+};
+
+/// Computes a volume version of Bondi radius of the worldtube \f$R\f$ from its
+/// boundary value (by repeating it over the radial dimension)
```

Comment:
for my own education, what is the `repeating` doing mathematically? I'm a bit confused by what is happening

Reply:
here we are creating radially-independent volume quantities from boundary quantities defined on the sphere, like $f(theta, phi, r) = F(theta, phi)$

Reply:
Okay, thank you for explaining that!